### PR TITLE
fix marginal_counts (which failed on pulse backend) (issue #6430)

### LINKED ISSUE #6430: marginal_counts is broken for pulse results
### Information

- **Qiskit Terra version**: 0.18.0
- **Python version**: 3.9
- **Operating system**: OSx

### What is the current behavior?
`marginal_counts` is broken for pulse jobs. I

### Steps to reproduce the problem
```python
from qiskit import pulse, QuantumCircuit, schedule, transpile, execute, IBMQ
from qiskit.result import marginal_counts


IBMQ.load_account()
provider = IBMQ.get_provider(hub='ibm-q', group='open', project='main')
backend = provider.get_backend('ibmq_armonk')

qc = QuantumCircuit(1,1)
qc.measure(0, 0)

job = backend.run(schedule(transpile(qc, backend), backend))
marginal_counts(job.result(), indices=[0])
```
Gives output
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-8e897d2c2a54> in <module>
     11 
     12 job = backend.run(schedule(transpile(qc, backend), backend))
---> 13 marginal_counts(job.result(), indices=[0])

~/opt/anaconda3/envs/python39/lib/python3.9/site-packages/qiskit/result/utils.py in marginal_counts(result, indices, inplace, format_marginal)
     55             experiment_result.data.counts = new_counts_hex
     56             experiment_result.header.memory_slots = len(indices)
---> 57             csize = experiment_result.header.creg_sizes
     58             experiment_result.header.creg_sizes = _adjust_creg_sizes(csize, indices)
     59         return result

AttributeError: 'QobjExperimentHeader' object has no attribute 'creg_sizes'
```

### Suggested solutions
It seems like it is assuming `creg` will be present in the header. This is not true for pulse jobs which do not have classical registers and therefore this field won't be present. The fix should make marginal_counts work when the header is empty.

### PR DESCRIPTION
**Pull request by chetmurthy**
Co-authored-by: rithikaadiga <56549493+rithikaadiga@users.noreply.github.com

When marginal_counts is passed a result that has ExperimentResults in it with headers that lack creg_sizes instance-variables, it just skips the "_adjust_creg_sizes" updating logic.

The test "test_result.py" is modified (by introducing a new test-class that subclasses the existing test) so that all the tests are run with versions of Result that have that field removed.  Again: there's a test there.  So subclass it, and in the subclass, during generation of Result-objects, remove the creg_sizes field.  Then all the tests will run (in the subclass).

### Details and comments

Fixes #6430

### CODE REVIEW

**Comment by chetmurthy** [comment]
Re: CLA, this happened with my PR for qiskit-machine-learning; I'm not sure why it doesn't recognize me as being a Github user (I mean, I'm logged-into Github, after all).  Sigh.

**Comment by 1ucian0** [comment]
Hi @chetmurthy, thanks a lot for the PR! 

It's tradition to include the author you copied code from [as `Co-authored-by`](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors). Do you mind amend your commit message with it?

File: qiskit/result/utils.py [hunk 1]
```diff
@@ -75,8 +75,12 @@ def marginal_counts(
...
             if indices is not None:
                 experiment_result.header.memory_slots = len(indices)
-                csize = experiment_result.header.creg_sizes
-                experiment_result.header.creg_sizes = _adjust_creg_sizes(csize, indices)
+                if (
+                    hasattr(experiment_result.header, "creg_sizes")
+                    and experiment_result.header.creg_sizes is not None
+                ):
+                    csize = experiment_result.header.creg_sizes
+                    experiment_result.header.creg_sizes = _adjust_creg_sizes(csize, indices)
```
Comment (1ucian0) [inline qiskit/result/utils.py:83]:
```suggestion
                csize = getattr(experiment_result.header, "creg_sizes")
                if (csize is not None):
                    experiment_result.header.creg_sizes = _adjust_creg_sizes(csize, indices)
```

File: releasenotes/fix-marginal_counts-on-pulse-backend.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    The function qiskit.result.marginal_counts(), when passed a Result from a
```
Comment (1ucian0) [inline releasenotes/fix-marginal_counts-on-pulse-backend.yaml:4]:
```suggestion
    The method :meth:`qiskit.result.marginal_counts`, when passed a :class"`.Result` from a
```

File: releasenotes/fix-marginal_counts-on-pulse-backend.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    pulse backend, would fail, b/c: such a result contains an array of
```
Comment (1ucian0) [inline releasenotes/fix-marginal_counts-on-pulse-backend.yaml:5]:
```suggestion
    pulse backend, would fail, because it contains an array of
```

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
checkout how we usually refer to code elements in the release notes.

**Comment by chetmurthy** [comment]
> Hi @chetmurthy, thanks a lot for the PR!
> 
> It's tradition to include the author you copied code from [as `Co-authored-by`](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors). Do you mind amend your commit message with it?

I can't modify the original commit without redoing the entire branch, so I added to co-authored-by to the PR description.  Will that work?

**Comment by chetmurthy** [comment]
Luciano, I think I resolved everything, yes?  Just checking.

File: releasenotes/fix-marginal_counts-on-pulse-backend.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    The method :meth:`qiskit.result.marginal_counts`, when passed a :class"`.Result` from a
```
Comment (1ucian0) [inline releasenotes/fix-marginal_counts-on-pulse-backend.yaml:4]:
```suggestion
    The method :meth:`qiskit.result.marginal_counts`, when passed a :class:`.Result` from a
```

File: test/python/result/test_result.py [hunk 5]
```diff
@@ -731,3 +731,19 @@ def test_memory_int_out(self):
...
+
+
+class TestResultOperationsNoCreg(TestResultOperations):
+    """Result operations methods, but without creg_sizes."""
+
+    def generate_qiskit_result(self):
+        """Generate Result (without creg_sizes for testing"""
+        rv = super().generate_qiskit_result()
+        del rv.results[0].header.creg_sizes
+        return rv
```
Comment (1ucian0) [inline test/python/result/test_result.py:749]: I think this test is a bit artificial on one side, and covers too much on the other one.

Comment (1ucian0) [reply test/python/result/test_result.py:749]: Here is my suggestion, closer to the original issue.

```suggestion
    def test_marginal_counts_no_cregs(self):
        """Test that marginal_counts without cregs See qiskit-terra/6430."""
        raw_counts_1 = {"0x0": 4, "0x1": 7, "0x2": 10, "0x6": 5, "0x9": 11, "0xD": 9, "0x12": 8}
        data_1 = models.ExperimentResultData(counts=dict(**raw_counts_1))
        exp_result_header_1 = QobjExperimentHeader(memory_slots=5)
        exp_result_1 = models.ExperimentResult(
            shots=54, success=True, data=data_1, header=exp_result_header_1
        )

        result = Result(results=[exp_result_1], **self.base_result_args)

        marginal_counts_result = marginal_counts(result, indices=[0])
```

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
I'm not fully sure about the testing. Have a look to my suggestion if that makes sense to you.

File: test/python/result/test_result.py [hunk 6]
```diff
@@ -731,3 +731,16 @@ def test_memory_int_out(self):
...
+        raw_counts_1 = {"0x0": 4, "0x1": 7, "0x2": 10, "0x6": 5, "0x9": 11, "0xD": 9, "0x12": 8}
+        data_1 = models.ExperimentResultData(counts=dict(**raw_counts_1))
+        exp_result_header_1 = QobjExperimentHeader(memory_slots=5)
+        exp_result_1 = models.ExperimentResult(
+            shots=54, success=True, data=data_1, header=exp_result_header_1
+        )
+
+        result = Result(results=[exp_result_1], **self.base_result_args)
+
+        _ = marginal_counts(result, indices=[0])
```
Comment (1ucian0) [inline test/python/result/test_result.py:746]: once the result is computed, asserting it is relatively cheap. You would check that also works-as-expected, in addition to it-works.
```suggestion
        _ = marginal_counts(result, indices=[0])
        marginal_counts_result = marginal_counts(result, indices=[0])
        self.assertEqual(marginal_counts_result.get_counts(), {"0": 27, "1": 27})
```

**Review by 1ucian0 — APPROVED** [review]
LGTM, thanks a lot for your work!

**Pull request merged**